arrow-libs/fx/arrow-fx-coroutines/src/commonMain/kotlin/arrow/fx/coroutines/CyclicBarrier.kt

```diff
@@ -23,28 +24,64 @@ public class CyclicBarrier(public val capacity: Int) {
       "Cyclic barrier must be constructed with positive non-zero capacity $capacity but was $capacity > 0"
     }
   }
-  
-  private data class State(val awaiting: Int, val epoch: Long, val unblock: CompletableDeferred<Unit>)
-  
-  private val state: AtomicRef<State> = AtomicRef(State(capacity, 0, CompletableDeferred()))
-  
+
+  private data class State(
+    /** The number of parties that are still required to join before the barrier will open and reset. **/
+    val awaiting: Int,
+    val epoch: Long,
+    val reset: Boolean,
+    val unblock: CompletableDeferred<Unit>,
+    val resetComplete: CompletableDeferred<Unit>
+  )
+
+  private val state: AtomicRef<State> =
+    AtomicRef(State(capacity, 0, false, CompletableDeferred(), CompletableDeferred()))
+
+  /** The number of parties currently waiting. **/
+  public val numberWaiting: Int
+    get() = capacity - state.get().awaiting
+
+  /**
+   * When called, all waiting coroutines will be cancelled with [CancellationException].
+   * When all threads have been cancelled the barrier will cycle.
+   */
+  public suspend fun reset() {
```

Comment:
I'm not happy that this method is a suspend function, and I would like to be able to remove this, using a similar logic to what currently exists to cycle the barrier when capacity is reached. My first attempt at doing that failed, however, as coroutine resumption seems to get in the way.

Further explanation - my implementation was such that I check `if (reset && awaiting == capacity) then state.update(s.copy(reset = false))`.  This doesn't work because the first coroutine to resume after `reset = true` is the oldest coroutine, which (I believe) has a view of state that is empty, so it immediately sets `reset = false`. Other coroutines then never see the reset.

Happy to be told I have no idea what I'm saying! I'm a beginner to coroutine concurrency.

Reply:
Resolved with [89457f5](https://github.com/arrow-kt/arrow/pull/3055/commits/89457f5bd5c34ad21476fa024c4e95b0a8631475).